tools/thanosconvert/thanosconvert.go

```diff
@@ -0,0 +1,188 @@
+package thanosconvert
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/objstore"
+
+	"github.com/cortexproject/cortex/pkg/storage/bucket"
+	cortex_tsdb "github.com/cortexproject/cortex/pkg/storage/tsdb"
+)
+
+// ThanosBlockConverter converts blocks written by Thanos to make them readable by Cortex
+type ThanosBlockConverter struct {
+	logger log.Logger
+	bkt    objstore.Bucket
+	dryRun bool
+}
+
+type PerUserResults struct {
+	FailedBlocks, ConvertedBlocks, UnchangedBlocks []string
+}
+
+func NewPerUserResults() PerUserResults {
+	return PerUserResults{
+		FailedBlocks:    []string{},
+		ConvertedBlocks: []string{},
+		UnchangedBlocks: []string{},
+	}
+}
+
+func (r *PerUserResults) AddFailed(blockID string) {
+	r.FailedBlocks = append(r.FailedBlocks, blockID)
+}
+func (r *PerUserResults) AddConverted(blockID string) {
+	r.ConvertedBlocks = append(r.ConvertedBlocks, blockID)
+}
+
+func (r *PerUserResults) AddUnchanged(blockID string) {
+	r.UnchangedBlocks = append(r.UnchangedBlocks, blockID)
+}
+
+type Results map[string]PerUserResults
+
+// NewThanosBlockConverter creates a ThanosBlockConverter
+func NewThanosBlockConverter(ctx context.Context, cfg bucket.Config, dryRun bool, logger log.Logger) (*ThanosBlockConverter, error) {
+	bkt, err := bucket.NewClient(ctx, cfg, "thanosconvert", logger, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ThanosBlockConverter{
+		bkt:    bkt,
+		logger: logger,
+		dryRun: dryRun,
+	}, err
+}
+
+// Run iterates over all blocks from all users in a bucket
+func (c ThanosBlockConverter) Run(ctx context.Context) (Results, error) {
+	results := make(Results)
+
+	scanner := cortex_tsdb.NewUsersScanner(c.bkt, cortex_tsdb.AllUsers, c.logger)
+	users, _, err := scanner.ScanUsers(ctx)
+	if err != nil {
+		return results, errors.Wrap(err, "error while scanning users")
+	}
+	level.Info(c.logger).Log("msg", "Scanned tenants")
+
+	for _, u := range users {
+		r, err := c.convertUser(ctx, u)
+		results[u] = r
+		if err != nil {
+			return results, errors.Wrap(err, fmt.Sprintf("error converting user %s", u))
+		}
+	}
+	return results, nil
+}
+
+func (c ThanosBlockConverter) convertUser(ctx context.Context, user string) (PerUserResults, error) {
+	results := PerUserResults{}
```

Comment:
I'd suggest to simplify this code by not keeping `PerUserResults` at all. The tool could simply use logging to log converted blocks, progress and failures, and not retain anything in the memory after processing of the user is done. It's only printed anyway -- we could print it immediately.

Reply:
Yes, good point - in my head I was thinking of putting out more sophisticated reporting (perhaps saving to a file), but this is probably unnecessary. On the other hand, this makes testing significantly easier, as we can run assertions on the contents of the results

Reply:
Good point about testing.